### ISSUE TITLE
Backtesting fails with Coins got no leverage tiers available

### ISSUE THREAD

**Issue by andreizolotarev** [freqtrade/freqtrade#11083]
## Describe your environment

  * Operating system: Linux Ubuntu
  * Python Version: 3.12.7 (`python -V`)
  * CCXT version: 4.4.24 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2024.10 (`freqtrade -V` or `docker compose run --rm freqtrade -V` for Freqtrade running in docker)
  
## Your question
Today I got an error during a backtesting (I'm using Bybit): 
![Image](https://github.com/user-attachments/assets/e5540004-56c4-48ff-97ff-2d2da1499852)
Although this has never happened before with these coins

Also I run ccxt directly and got no leverage tiers for the coins (SUI is missing):
![Image](https://github.com/user-attachments/assets/1c8e92cf-2792-466f-83dd-217124162aa9)
at the same time Bybit returns SUI leverage https://api-testnet.bybit.com/v5/market/risk-limit?category=linear&symbol=SUIUSDT

Can you tell me what could be wrong?

The config and strategy to reproduce:
```json
{
    "$schema": "https://schema.freqtrade.io/schema.json",
    "max_open_trades": 10,
    "stake_currency": "USDT",
    "stake_amount": "unlimited",
    "tradable_balance_ratio": 0.99,
    "fiat_display_currency": "USD",
    "dry_run": true,
    "dry_run_wallet": 1000,
    "cancel_open_orders_on_exit": false,
    "trading_mode": "futures",
    "margin_mode": "isolated",
    "unfilledtimeout": {
        "entry": 10,
        "exit": 10,
        "exit_timeout_count": 0,
        "unit": "minutes"
    },
    "entry_pricing": {
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1,
        "price_last_balance": 0.0,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "exit_pricing":{
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1
    },
    "exchange": {
        "name": "bybit",
        "key": "",
        "secret": "",
        "ccxt_config": {},
        "ccxt_async_config": {},
        "pair_whitelist": [
            "BTC/USDT:USDT",
            "ETH/USDT:USDT",
            "XRP/USDT:USDT",
            "DOGE/USDT:USDT",
            "SOL/USDT:USDT",
            "1000PEPE/USDT:USDT",
            "SUI/USDT:USDT",
            "LINK/USDT:USDT",
            "ADA/USDT:USDT",
            "AAVE/USDT:USDT",
            "PNUT/USDT:USDT", 
            "WIF/USDT:USDT",
            "AVAX/USDT:USDT",
            "ONDO/USDT:USDT",
            "GOAT/USDT:USDT",
            "HBAR/USDT:USDT",
            "ENA/USDT:USDT",
            "SHIB1000/USDT:USDT",
            "FTM/USDT:USDT",
            "GALA/USDT:USDT",
            "1000000BABYDOGE/USDT",
            "LTC/USDT:USDT",
            "SAND/USDT:USDT",
            "DOT/USDT:USDT",
            "MOVE/USDT:USDT",
            "UNI/USDT:USDT",
            "SUSHI/USDT:USDT",
            "ENS/USDT:USDT",
            "NEAR/USDT:USDT",
            "CRV/USDT:USDT",
            "APT/USDT:USDT",
            "1000BONK/USDT:USDT",
            "ORDI/USDT:USDT",
            "POPCAT/USDT:USDT",
            "XVG/USDT:USDT",
            "XLM/USDT:USDT",
            "SEI/USDT:USDT",
            "ARB/USDT:USDT",         
            "BNB/USDT:USDT",
            "TIA/USDT:USDT",
            "OP/USDT:USDT",
            "ME/USDT:USDT",
            "ETC/USDT:USDT",
            "JASMY/USDT:USDT",
            "ALGO/USDT:USDT",
            "MOODENG/USDT:USDT",
            "VIRTUAL/USDT:USDT",
            "EIGEN/USDT:USDT",
            "1000FLOKI/USDT:USDT",
            "BCH/USDT:USDT",
            "INJ/USDT:USDT",
            "STX/USDT:USDT",
            "IO/USDT:USDT",
            "FIL/USDT:USDT",
            "IOTA/USDT:USDT",
            "OL/USDT:USDT",
            "GRASS/USDT:USDT",
            "CVC/USDT:USDT",
            "RENDER/USDT:USDT",
            "VET/USDT:USDT",
            "TON/USDT:USDT",
            "BRETT/USDT:USDT",
            "ACT/USDT:USDT",
            "KAS/USDT:USDT",
            "APE/USDT:USDT",
            "JUP/USDT:USDT",
            "CTC/USDT:USDT",
            "MNT/USDT:USDT",
            "DEEP/USDT:USDT",
            "POL/USDT:USDT",
            "MORPHO/USDT:USDT",
            "DRIFT/USDT:USDT",
            "ZRX/USDT:USDT",
            "LDO/USDT:USDT",
            "ATOM/USDT:USDT",
            "TRX/USDT:USDT"
        ],
        "pair_blacklist": [
            "USDT/USDC:USDT",
            "WLD/USDT:USDT",
            "EOS/USDT:USDT",
            "MEMEFI/USDT:USDT",
            "ICP/USDT:USDT",
            "CHILLGUY/USDT:USDT",
            "AERGO/USDT:USDT",
            "JTO/USDT:USDT",
            "BEAM/USDT:USDT",
            "TAO/USDT:USDT",
            "DYDX/USDT:USDT"
        ]
    },
    "pairlists": [
        {
            "method": "StaticPairList"
        }
    ],
    "telegram": {
        "enabled": false,
        "token": "",
        "chat_id": ""
    },
    "api_server": {
        "enabled": false,
        "listen_ip_address": "127.0.0.1",
        "listen_port": 8080,
        "verbosity": "error",
        "enable_openapi": false,
        "jwt_secret_key": "6483211e574bf331624b5b61cce2930eb455e71153bf79a54d195ca4da2bcfd4",
        "ws_token": "tV4Vm-ZAi9KhIDdrNsKmTQPYG_A6fXkQ5A",
        "CORS_origins": [],
        "username": "",
        "password": ""
    },
    "bot_name": "freqtrade",
    "initial_state": "running",
    "force_entry_enable": false,
    "internals": {
        "process_throttle_secs": 5
    }
}
```
```python
# pragma pylint: disable=missing-docstring, invalid-name, pointless-string-statement
# flake8: noqa: F401
# isort: skip_file
# --- Do not remove these imports ---
import numpy as np
import pandas as pd
from datetime import datetime, timedelta, timezone
from pandas import DataFrame
from typing import Optional, Union

from freqtrade.strategy import (
    IStrategy,
    Trade,
    Order,
    PairLocks,
    informative,  # @informative decorator
    # Hyperopt Parameters
    BooleanParameter,
    CategoricalParameter,
    DecimalParameter,
    IntParameter,
    RealParameter,
    # timeframe helpers
    timeframe_to_minutes,
    timeframe_to_next_date,
    timeframe_to_prev_date,
    # Strategy helper functions
    merge_informative_pair,
    stoploss_from_absolute,
    stoploss_from_open,
)

# --------------------------------
# Add your lib to import here
import talib.abstract as ta
from technical import qtpylib


# This class is a sample. Feel free to customize it.
class SampleStrategy(IStrategy):
    """
    This is a sample strategy to inspire you.
    More information in https://www.freqtrade.io/en/latest/strategy-customization/

    You can:
        :return: a Dataframe with all mandatory indicators for the strategies
    - Rename the class name (Do not forget to update class_name)
    - Add any methods you want to build your strategy
    - Add any lib you need to build your strategy

    You must keep:
    - the lib in the section "Do not remove these libs"
    - the methods: populate_indicators, populate_entry_trend, populate_exit_trend
    You should keep:
    - timeframe, minimal_roi, stoploss, trailing_*
    """

    # Strategy interface version - allow new iterations of the strategy interface.
    # Check the documentation or the Sample strategy to get the latest version.
    INTERFACE_VERSION = 3

    # Can this strategy go short?
    can_short: bool = True

    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi".
    minimal_roi = {
        # "120": 0.0,  # exit after 120 minutes at break even
        "60": 0.01,
        "30": 0.02,
        "0": 0.04,
    }

    # Optimal stoploss designed for the strategy.
    # This attribute will be overridden if the config file contains "stoploss".
    stoploss = -0.10

    # Trailing stoploss
    trailing_stop = False
    # trailing_only_offset_is_reached = False
    # trailing_stop_positive = 0.01
    # trailing_stop_positive_offset = 0.0  # Disabled / not configured

    # Optimal timeframe for the strategy.
    timeframe = "1d"

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = True

    # These values can be overridden in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False

    # Hyperoptable parameters
    buy_rsi = IntParameter(low=1, high=50, default=30, space="buy", optimize=True, load=True)
    sell_rsi = IntParameter(low=50, high=100, default=70, space="sell", optimize=True, load=True)
    short_rsi = IntParameter(low=51, high=100, default=70, space="sell", optimize=True, load=True)
    exit_short_rsi = IntParameter(low=1, high=50, default=30, space="buy", optimize=True, load=True)

    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 200

    # Optional order type mapping.
    order_types = {
        "entry": "limit",
        "exit": "limit",
        "stoploss": "market",
        "stoploss_on_exchange": False,
    }

    # Optional order time in force.
    order_time_in_force = {"entry": "GTC", "exit": "GTC"}

    plot_config = {
        "main_plot": {
            "tema": {},
            "sar": {"color": "white"},
        },
        "subplots": {
            "MACD": {
                "macd": {"color": "blue"},
                "macdsignal": {"color": "orange"},
            },
            "RSI": {
                "rsi": {"color": "red"},
            },
        },
    }

    def informative_pairs(self):

        return []

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        # ADX
        dataframe["adx"] = ta.ADX(dataframe)

        # RSI
        dataframe["rsi"] = ta.RSI(dataframe)

        # Stochastic Fast
        stoch_fast = ta.STOCHF(dataframe)
        dataframe["fastd"] = stoch_fast["fastd"]
        dataframe["fastk"] = stoch_fast["fastk"]

        # MACD
        macd = ta.MACD(dataframe)
        dataframe["macd"] = macd["macd"]
        dataframe["macdsignal"] = macd["macdsignal"]
        dataframe["macdhist"] = macd["macdhist"]

        # MFI
        dataframe["mfi"] = ta.MFI(dataframe)

        # Bollinger Bands
        bollinger = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe["bb_lowerband"] = bollinger["lower"]
        dataframe["bb_middleband"] = bollinger["mid"]
        dataframe["bb_upperband"] = bollinger["upper"]
        dataframe["bb_percent"] = (dataframe["close"] - dataframe["bb_lowerband"]) / (
            dataframe["bb_upperband"] - dataframe["bb_lowerband"]
        )
        dataframe["bb_width"] = (dataframe["bb_upperband"] - dataframe["bb_lowerband"]) / dataframe[
            "bb_middleband"
        ]

        dataframe["sar"] = ta.SAR(dataframe)

        dataframe["tema"] = ta.TEMA(dataframe, timeperiod=9)

        hilbert = ta.HT_SINE(dataframe)
        dataframe["htsine"] = hilbert["sine"]
        dataframe["htleadsine"] = hilbert["leadsine"]

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the entry signal for the given dataframe
        :param dataframe: DataFrame
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with entry columns populated
        """
        dataframe.loc[
            (
                # Signal: RSI crosses above 30
                (qtpylib.crossed_above(dataframe["rsi"], self.buy_rsi.value))
                & (dataframe["tema"] <= dataframe["bb_middleband"])  # Guard: tema below BB middle
                & (dataframe["tema"] > dataframe["tema"].shift(1))  # Guard: tema is raising
                & (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "enter_long",
        ] = 1

        dataframe.loc[
            (
                # Signal: RSI crosses above 70
                (qtpylib.crossed_above(dataframe["rsi"], self.short_rsi.value))
                & (dataframe["tema"] > dataframe["bb_middleband"])  # Guard: tema above BB middle
                & (dataframe["tema"] < dataframe["tema"].shift(1))  # Guard: tema is falling
                & (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "enter_short",
        ] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the exit signal for the given dataframe
        :param dataframe: DataFrame
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with exit columns populated
        """
        dataframe.loc[
            (
                # Signal: RSI crosses above 70
                (qtpylib.crossed_above(dataframe["rsi"], self.sell_rsi.value))
                & (dataframe["tema"] > dataframe["bb_middleband"])  # Guard: tema above BB middle
                & (dataframe["tema"] < dataframe["tema"].shift(1))  # Guard: tema is falling
                & (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "exit_long",
        ] = 1

        dataframe.loc[
            (
                # Signal: RSI crosses above 30
                (qtpylib.crossed_above(dataframe["rsi"], self.exit_short_rsi.value))
                &
                # Guard: tema below BB middle
                (dataframe["tema"] <= dataframe["bb_middleband"])
                & (dataframe["tema"] > dataframe["tema"].shift(1))  # Guard: tema is raising
                & (dataframe["volume"] > 0)  # Make sure Volume is not 0
            ),
            "exit_short",
        ] = 1

        return dataframe

```
*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

**Comment by xmatthias** [comment]
To avoid fetching leverage tiers over and over again, freqtrade does cache leverage tiers.
For binance - it's hard-cached by us (as it requires credentials to access) and updated weekly (develop version).

For other exchanges, it's downloaded on first start, and usually cached for a few weeks.
To void the caching, simply delete the caching file (`user_data/data/bybit/futures/leverage_tiers_{stake_currency}.json` (make sure to update the path if you're using a different data directory).
Also - {stake_currency} is obviously USDT in your case - but could just as well be USDC or whatever other stable coin the exchange supports.

**Comment by andreizolotarev** [comment]
I deleted the caching file and also got the same error during backtesting. Redownloading data with empty data folder also didn't help.

I run it in debug mode. 
```python
exchange = ccxt.bybit(
    {'options':
        {'trading_mode': 'futures',
         "margin_mode": "isolated"}
    }
)
res = exchange.fetch_leverage_tiers()
``` 
Here's what I found, maybe it will help someone else avoid this error in backtesting. 

In the ccxt's `fetch_paginated_call_cursor` method there is a variable `maxCalls`, it was equal to 20 and it seems that this number of calls/attempts was not enough to get all coins. If you increase this value (I reassigned it to 50), then the leverage information will be received for more coins and the error will disappear.

As you can see after this change SUI is in the response and the backtesting passed without errors
![Image](https://github.com/user-attachments/assets/05f536da-c54d-44eb-9d45-5a94d827cf5b)

**Comment by xmatthias** [comment]
Nice spot - seems to be hardcoded in ccxt. 
Opened an issue for this against ccxt https://github.com/ccxt/ccxt/issues/24549

**Comment by LuisAlejandro** [comment]
btw, just as a comment if anyone bumps into this problem in August 2025: the fix for this is going to be released in Freqtrade `2025.8`

fix got introduced in this PR last week: https://github.com/freqtrade/freqtrade/pull/12091

**Comment by AchernarAW** [comment]
> 顺便说一句，如果有人在 2025 年 8 月遇到这个问题，请发表评论：此修复程序将在 Freqtrade 中发布`2025.8`
> 
> 上周在此 PR 中引入了修复程序：[#12091](https://github.com/freqtrade/freqtrade/pull/12091)

<img width="2141" height="487" alt="Image" src="https://github.com/user-attachments/assets/25d043f4-e046-4f1f-862e-94170522f6d6" />

I use the Gate exchange and am currently experiencing this issue as well, using the latest develop version.

**Comment by xmatthias** [comment]
@AchernarAW you're then either having problems with your network - or you're in a jurisdiction where gate responses differ.
If you have a problem, best open a new issue, filling out the issue template, and including your full logs (as text, not as screenshot).

It's at any rate not related to this issue - which was fixed in 2025.8, i'll hence close this issue to avoid further notification spam to people subscribed here.
